CLI: Support lower-case test argument passing

okurz opened the pull request:
Motivation:
isotovideo from os-autoinst automatically transforms lower-case test
parameter command line arguments into upper-case test arguments. openQA
so far only checked the validity of upper-case parameters in a regex,
failing if lower-case was used.

Design Choices:
We optionally transform lower-case test parameter keys into upper-case
test arguments during `openqa-cli schedule`. The uppercasing applies
only to parameter names (keys), not values, and skips special API keys
like `async` and `scheduled_product_clone_id` which have been
deduplicated. Other commands like `openqa-cli api` continue to
preserve precise casing.

User Benefits:
Users can now conveniently pass lower-case test arguments via
`openqa-cli schedule` without encountering errors, aligning the
experience with `isotovideo`.

okurz committed:
feat(cli): support lower-case test argument passing

Motivation:
isotovideo from os-autoinst automatically transforms lower-case test
parameter command line arguments into upper-case test arguments. openQA
so far only checked the validity of upper-case parameters in a regex,
failing if lower-case was used.

Design Choices:
We optionally transform lower-case test parameter keys into upper-case
test arguments during `openqa-cli schedule`. The uppercasing applies
only to parameter names (keys), not values, and skips special API keys
like `async` and `scheduled_product_clone_id` which have been
deduplicated. Other commands like `openqa-cli api` continue to
preserve precise casing.

User Benefits:
Users can now conveniently pass lower-case test arguments via
`openqa-cli schedule` without encountering errors, aligning the
experience with `isotovideo`.

Martchus reviewed (changes requested):
Wouldn't this upper-case any kind of parameter (and not only ones ending up job settings)? Then it would be very wrong.

okurz commented:
> Wouldn't this upper-case any kind of parameter (and not only ones ending up job settings)? Then it would be very wrong.

Yes, I agree :) I haven't checked that. Tests should tell but apparently my working copy of openQA is outdated.

@mergifyio rebase

perlpunk commented:
> Wouldn't this upper-case any kind of parameter (and not only ones ending up job settings)? Then it would be very wrong.

Yeah, and it would also uppercase the value of the setting, e.g. `distri=Tumbleweed` => `DISTRI=TUMBLEWEED`

okurz commented:
just rebased, not more

okurz commented:
> Wouldn't this upper-case any kind of parameter (and not only ones ending up job settings)? Then it would be very wrong.

Fixed this by excluding special parameters

> Yeah, and it would also uppercase the value of the setting, e.g. `distri=Tumbleweed` => `DISTRI=TUMBLEWEED`

Fixed by only upcasing the key

Martchus reviewed (dismissed):
I would implement this in the controller code where we handle such API requests instead of making the case-insensitive behavior client-specific. This would avoid the problems mentioned in inline comments.

okurz commented on lib/OpenQA/Command.pm:
```diff
@@ -16,6 +16,7 @@ use Term::ANSIColor qw(colored);
 my $JSON = Cpanel::JSON::XS->new->utf8->canonical->allow_nonref->allow_unknown->allow_blessed->convert_blessed
   ->stringify_infnan->escape_slash->allow_dupkeys->pretty;
 my $PARAM_RE = qr/^([[:alnum:]_\[\]\.\:]+)=(.*)$/s;
+my $RESERVED_API_KEYS_RE = qr/^(async|scheduled_product_clone_id)$/;
```
I now moved the functionality into lib/OpenQA/WebAPI/Controller/API/V1/Iso.pm `create` where we already called `uc`. There we know that this list actually *is* exhaustive as no other parameters had been extracted before calling `uc`.

okurz commented on lib/OpenQA/Command.pm:
```diff
@@ -79,16 +80,20 @@ sub parse_headers ($self, @headers) {
 }
 
-sub parse_params ($self, $args, $param_file) {
+sub parse_params ($self, $args, $param_file, $uc_keys = 0) {
```
I now moved the functionality into lib/OpenQA/WebAPI/Controller/API/V1/Iso.pm `create` where we already called `uc` which should means that we handle the transformation more low level so should ensure consistent behaviour.

okurz commented on t/43-cli-schedule.t:
```diff
@@ -106,6 +107,11 @@ subtest 'scheduling and monitoring set of two jobs' => sub {
+    combined_like { $res = $schedule->run(@options, @scenarios, @settings4) }
+    qr|2 jobs have been created.*(http://127.0.0.1.*/tests/\d+.*){2}passed.*softfailed|s,
+      'response logged if jobs created correctly from lower-case arguments';
+    is $res, 0, 'zero return-code if all jobs are ok with lower-case arguments';
```
done

okurz commented:
incorporated review suggestions from @Martchus

Martchus commented on t/43-cli-schedule.t:
```diff
@@ -132,4 +139,16 @@ subtest 'monitor jobs as a separate command' => sub {
     is $res, 0, 'zero return-code if clone of followed jobs ok';
 };
 
+subtest 'api command preserves lowercase keys for routes expecting them' => sub {
+    my $res;
+    my $api = OpenQA::CLI::api->new;
+    my @api_options = ('--apikey', 'ARTHURKEY01', '--apisecret', 'EXCALIBUR', '--host', $host);
+    combined_like {
+        $res = $api->run(@api_options, 'isos/job_stats', 'distri=example', 'version=0', 'flavor=DVD', 'arch=x86_64',
+            'build=123')
+    }
+    qr/\{\}/s, 'response logged successfully because lower-case arguments were preserved';
+    is $res, 0, 'zero return-code for successful api call with lower-case arguments';
+};
+
```
Considering you're now not touching `Command.pm` anymore (except for adding a blank line), this test is probably not needed anymore. Of course it can't hurt so you can also keep it.

okurz replied:
true. I would prefer to keep it as it nicely showcases how one can use the CLI

Martchus commented on lib/OpenQA/Command.pm:
```diff
@@ -83,6 +83,7 @@ sub parse_headers ($self, @headers) {
 
 sub parse_params ($self, $args, $param_file) {
     my %params;
+
```
```suggestion
```

okurz replied:
done

okurz commented:
incorporated the suggestion from @Martchus and rebased
